runtime/compiler/optimizer/J9ValuePropagation.cpp

```diff
@@ -639,6 +639,165 @@ J9::ValuePropagation::constrainRecognizedMethod(TR::Node *node)
       {
       switch (rm)
          {
+         case TR::java_lang_invoke_MethodHandle_asType:
+            {
+            TR::Node* mh = node->getChild(0);
+            TR::Node* mt = node->getChild(1);
+            bool mhConstraintGlobal, mtConstraintGlobal;
+            TR::VPConstraint* mhConstraint = getConstraint(mh, mhConstraintGlobal);
+            TR::VPConstraint* mtConstraint = getConstraint(mt, mtConstraintGlobal);
+            if (mhConstraint
+                && mtConstraint
+                && mhConstraint->getKnownObject()
+                && mhConstraint->isNonNullObject()
+                && mtConstraint->getKnownObject()
+                && mtConstraint->isNonNullObject())
+               {
+               uintptrj_t* mhLocation = getObjectLocationFromConstraint(mhConstraint);
+               uintptrj_t* mtLocation = getObjectLocationFromConstraint(mtConstraint);
+               uint32_t mtOffset = J9VMJAVALANGINVOKEMETHODHANDLE_TYPE_OFFSET(comp()->fej9()->vmThread());
+
+               TR::VMAccessCriticalSection asType(comp(),
+                                                           TR::VMAccessCriticalSection::tryToAcquireVMAccess);
+               if (!asType.hasVMAccess())
+                  break;
+               uintptrj_t mhObject = comp()->fej9()->getStaticReferenceFieldAtAddress((uintptrj_t)mhLocation);
+               uintptrj_t mtInMh = comp()->fej9()->getReferenceFieldAtAddress(mhObject + mtOffset);
+               uintptrj_t mtObject = comp()->fej9()->getStaticReferenceFieldAtAddress((uintptrj_t)mtLocation);
+
+               if (trace())
+                  traceMsg(comp(), "mhLocation %p mtLocation %p mhObject %p mtOffset %d mtInMh %p mtObject %p\n", mhLocation, mtLocation, mhObject, mtOffset, mtInMh, mtObject);
+
+               // AOT doesn't relocate/validate objects, which is why this transformation is disabled under AOT
```

Comment:
Where's the code for this comment? I don't see anything obvious about AOT here

Reply:
There is an if statement right before the the switch statement to bail out if we're in AOT. There are two switch statements in this function, one for transformations that work in AOT, one for transformations that won't work in AOT. I can merge the two switch statements and add the AOT check before optimizations that don't have AOT support.